datastore/protocol.go

```diff
@@ -1,163 +1,245 @@
+		access_password numeric(20),
+		update_password numeric(20),
+		reference_count numeric(10) NOT NULL DEFAULT 0,
+		creation_date timestamp DEFAULT (CURRENT_TIMESTAMP AT TIME ZONE 'UTC'),
+		update_date timestamp DEFAULT (CURRENT_TIMESTAMP AT TIME ZONE 'UTC'),
+		true_update_date timestamp DEFAULT (CURRENT_TIMESTAMP AT TIME ZONE 'UTC'),
+		last_reference_date timestamp DEFAULT (CURRENT_TIMESTAMP AT TIME ZONE 'UTC'),
+		expiration_date timestamp,
+		upload_completed boolean
+	)`)
+	if err != nil {
+		common_globals.Logger.Error(err.Error())
+		return
 	}
 
-	// * Allow anyone
-	if permission.Permission == 0 {
-		return nil
+	// * Object persistence is handled at the user level,
+	// * not the object level
+	_, err = manager.Database.Exec(`CREATE TABLE datastore.persistence_slots (
+		pid numeric(20),
+		slot int, -- can technically only be 0-15, but sent as a uint16
+		data_id numeric(20) REFERENCES datastore.objects(data_id),
+		PRIMARY KEY (pid, slot)
+	)`)
+	if err != nil {
+		common_globals.Logger.Error(err.Error())
+		return
 	}
 
-	// * Allow only friends of the owner
-	if permission.Permission == 1 {
-		// TODO - This assumes a legacy client. Will not work on the Switch
-		friendsList := c.GetUserFriendPIDs(uint32(ownerPID))
-
-		if !slices.Contains(friendsList, uint32(accessorPID)) {
-			return nex.NewError(nex.ResultCodes.DataStore.PermissionDenied, "change_error")
-		}
+	_, err = manager.Database.Exec(`CREATE TABLE datastore.rating_settings (
+		data_id numeric(20) NOT NULL REFERENCES datastore.objects(data_id),
+
+		-- Data set in DataStoreRatingInitParamWithSlot
+		slot smallint, -- can technically only be 0-15, but allocate enough space anyway
+		raw_flags smallint,
+		raw_internal_flags smallint,
+		minimum_value numeric(10),
+		maximum_value numeric(10),
+		initial_value numeric(20),
+		lock_type smallint, -- can technically only be 0-3, but allocate enough space anyway
+		lock_period_duration int,
+		lock_period_hour smallint, -- can technically only hold 2 digits, but allocate enough space anyway
+
+		-- Decoded raw_flags
+		-- Only supports stock flags, custom flags must be handled separately
+		allow_multiple_ratings boolean NOT NULL DEFAULT FALSE,
+		round_negatives boolean NOT NULL DEFAULT TRUE,
+		disable_self_rating boolean NOT NULL DEFAULT FALSE,
+
+		-- Decoded raw_internal_flags
+		-- Only supports stock flags, custom flags must be handled separately
+		use_minimum boolean NOT NULL DEFAULT FALSE,
+		use_maximum boolean NOT NULL DEFAULT FALSE,
+
+		PRIMARY KEY (data_id, slot)
+	);`)
+	if err != nil {
+		common_globals.Logger.Error(err.Error())
+		return
 	}
 
-	// * Allow only users whose PIDs are defined in permission.RecipientIDs
-	if permission.Permission == 2 {
-		if !permission.RecipientIDs.Contains(accessorPID) {
-			return nex.NewError(nex.ResultCodes.DataStore.PermissionDenied, "change_error")
-		}
+	_, err = manager.Database.Exec(`CREATE TABLE datastore.ratings (
+		id serial PRIMARY KEY,
```

Comment:
We should use bigserial here for good measure

```suggestion
		id bigserial PRIMARY KEY,
```